core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,243 @@
+      override def run(): Unit = {
+        Utils.logUncaughtExceptions(expireExecutorsInBlackList(executorIdToFailureStatus))
+      }
+    }
+    scheduler.scheduleAtFixedRate(scheduleTask, 0L, recoverPeriod, TimeUnit.SECONDS)
+  }
+
+  def stop(): Unit = {
+    scheduler.shutdown()
+    scheduler.awaitTermination(10, TimeUnit.SECONDS)
+  }
+
+  // The actual implementation is delegated to strategy
+  private def expireExecutorsInBlackList(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus]): Unit = synchronized {
+    strategy.expireExecutorsInBlackList(executorIdToFailureStatus)
+
+    invalidAllCache()
+  }
+
+  def updateFailureExecutors(info: TaskInfo, reason: TaskEndReason) : Unit = synchronized {
+    reason match {
+      // If task succeeding, remove related record from executorIdToFailureStatus
+      case Success =>
+        removeFailureExecutorsForTaskId(info.executorId, info.taskId)
+
+      // If task failing, update latest failure time and failedTaskIds
+      case _ =>
+        val executorId = info.executorId
+        executorIdToFailureStatus.get(executorId) match {
+          case Some(failureStatus) =>
+            failureStatus.updatedTime = clock.getTimeMillis()
+            val failedTimes = failureStatus.numFailuresPerTask.getOrElse(info.taskId, 0) + 1
+            failureStatus.numFailuresPerTask.update(info.taskId, failedTimes)
+          case None =>
+            val failedTasks = mutable.HashMap(info.taskId -> 1)
+            val failureStatus = new FailureStatus(
+              clock.getTimeMillis(),
+              info.host,
+              failedTasks)
+            executorIdToFailureStatus.update(executorId, failureStatus)
+        }
+        invalidAllCache()
+    }
+  }
+
+  // remove the executorId from executorIdToFailureStatus
+  def removeFailureExecutors(executorId: String) : Unit = synchronized {
+    executorIdToFailureStatus.remove(executorId)
+    invalidAllCache()
+  }
+
+  // remove the failure record related to given taskId from executorIdToFailureStatus. If the
+  // number of records of given executorId becomes 0, remove the completed executorId.
+  def removeFailureExecutorsForTaskId(
+      executorId: String,
+      taskId: Long) : Unit = synchronized {
+    executorIdToFailureStatus.get(executorId).map(fs => {
+      fs.numFailuresPerTask.remove(taskId)
+      if(fs.numFailuresPerTask.isEmpty){
+        executorIdToFailureStatus.remove(executorId)
+      }
+      invalidAllCache()
+    })
+  }
+
+  def executorIsBlacklisted(
+      executorId: String,
+      sched: TaskSchedulerImpl,
+      taskId: Long) : Boolean = {
+
+    executorBlacklist(sched, taskId).contains(executorId)
+  }
+
+  // If the node is in blacklist, all executors allocated on that node will
+  // also be put into  executor blacklist.
+  // By default it's turned off, user can enable it in sparkConf.
+  private def speculationFailedExecutor(
+      sched: TaskSchedulerImpl): Set[String] = {
```

Comment:
I am fine to make it always on and remove the configuration. It was there because I want to achieve the same logic as before. But I just noticed that according to current implementation, node blacklist will always be empty under "default" strategy and "speculation" is not useful. I will remove it. Thanks for the comment.
I